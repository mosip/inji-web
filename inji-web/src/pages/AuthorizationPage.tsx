import React, {useEffect, useState} from "react";
import {SpinningLoader} from "../components/Common/SpinningLoader";
import {api} from "../utils/api";
import {LandingPageWrapper} from "../components/Common/LandingPageWrapper";
import {ErrorSheildIcon} from "../components/Common/ErrorSheildIcon";
import {useTranslation} from "react-i18next";

export const AuthorizationPage: React.FC = () => {
    const url = window.location.href;
    const currentQueryParams = new URLSearchParams(window.location.search);
    const [error, setError] = useState<String>(currentQueryParams.get("error")+"");
    const {t} = useTranslation("AuthorizationPage");

    useEffect( () => {
        if(url.indexOf("error") === -1) {
            window.location.href = api.mimotoHost + "/authorize" + window.location.search;
        }
    },[])

    if(url.indexOf("error") == -1){
        return <div><LandingPageWrapper icon={<SpinningLoader/>} title={""} subTitle={""} gotoHome={false}/></div>
    }
    return <div><LandingPageWrapper icon={<ErrorSheildIcon />} title={t(`error.code.${error}`)} subTitle={t(`error.message.${error}`)} gotoHome={true}/></div>
}


