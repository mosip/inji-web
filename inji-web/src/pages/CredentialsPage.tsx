import React, {useEffect, useState} from "react";
import {useParams} from "react-router-dom";
import {RequestStatus, useFetch} from "../hooks/useFetch";
import {NavBar} from "../components/Common/NavBar";
import {CredentialList} from "../components/Credentials/CredentialList";
import {useDispatch, useSelector} from "react-redux";
import {storeSelectedIssuer} from "../redux/reducers/issuersReducer";
import {storeCredentials, storeFilteredCredentials} from "../redux/reducers/credentialsReducer";
import {api} from "../utils/api";
import {useTranslation} from "react-i18next";
import {toast} from "react-toastify";

import {ApiRequest, DisplayArrayObject, IssuerObject} from "../types/data";
import {getObjectForCurrentLanguage} from "../utils/i18n";
import {RootState} from "../types/redux";
import {isObjectEmpty} from "../utils/misc";

export const CredentialsPage: React.FC = () => {

    const {state, fetchRequest} = useFetch();
    const params = useParams<CredentialParamProps>();
    const dispatch = useDispatch();
    const {t} = useTranslation("CredentialsPage");
    const language = useSelector((state: RootState) => state.common.language);
    let displayObject= {} as DisplayArrayObject;
    let [selectedIssuer, setSelectedIssuer] = useState({} as IssuerObject)
    if(!isObjectEmpty(selectedIssuer)){
        displayObject = getObjectForCurrentLanguage(selectedIssuer.display, language);
    }

    useEffect(() => {
        const fetchCall = async () => {
            let apiRequest: ApiRequest = api.fetchSpecificIssuer;
            let response = await fetchRequest(
                apiRequest.url(params.issuerId ?? ""),
                apiRequest.methodType,
                apiRequest.headers()
            );
            dispatch(storeSelectedIssuer(response?.response));
            setSelectedIssuer(response?.response);

            apiRequest = api.fetchCredentialTypes2;
            response = await fetchRequest(
                apiRequest.url(params.issuerId ?? ""),
                apiRequest.methodType,
                apiRequest.headers()
            );
            // apiRequest = api.fetchCredentialTypes;
            // response = await fetchRequest(
            //     apiRequest.url(response?.response[".well-known"]),
            //     apiRequest.methodType,
            //     apiRequest.headers()
            // );
            dispatch(storeFilteredCredentials(response?.response?.supportedCredentials));
            dispatch(storeCredentials(response?.response?.supportedCredentials));
        }
        fetchCall();
    }, [])

    if (state === RequestStatus.ERROR) {
        toast.error(t("errorContent"));
    }


    return <React.Fragment>
        <div className="bg-iw-background min-h-screen"
             data-testid="Credentials-Page-Container">
            <NavBar title={displayObject?.name} search={true} fetchRequest={fetchRequest} link={"/"}/>
            <div data-testid="Credential-List-Container" className="container mx-auto mt-8 px-10 sm:px-0">
                <CredentialList state={state}/>
            </div>
        </div>
    </React.Fragment>
}

