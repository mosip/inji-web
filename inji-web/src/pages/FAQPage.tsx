import React from 'react';
import {FAQAccordion} from '../components/Faq/FAQAccordion';
import {useTranslation} from 'react-i18next';
import {useLocation, useNavigate} from 'react-router-dom';
import {FAQPageProps} from '../components/Dashboard/types';
import {navigateToDashboardHome} from './Dashboard/utils';
import {PageTitle} from "../components/Common/PageTitle/PageTitle";
import { TertiaryButton } from './Dashboard/TertiaryButton';

export const FAQPage: React.FC<FAQPageProps> = ({backUrl, withHome}) => {
    const {t} = useTranslation(['FAQPage', 'Dashboard']);
    const navigate = useNavigate();
    const location = useLocation();
    const previousPagePath = location.state?.from;

    const handleBackClick = () => {
        if (backUrl) {
            navigate(backUrl); // Navigate to the URL sent by the parent
        } else if (previousPagePath) {
            navigate(previousPagePath); // Navigate to the previous link in history
        } else {
            navigateToDashboardHome(navigate); // Navigate to homepage if opened directly
        }
    };

    return (
        <div className="w-full flex flex-col sm:flex-row justify-between items-start p-1 sm:p-2 md:p-4 flex-grow overflow-y-auto">
            <div className="flex flex-row items-start gap-4 w-full">
                <div className="flex items-start">
                    <svg
                        width="29"
                        height="29"
                        viewBox="0 0 24 18"
                        fill="none"
                        xmlns="http://www.w3.org/2000/svg"
                        className="mr-2 cursor-pointer"
                        onClick={handleBackClick}
                    >
                        <path
                            d="M20 11H7.83l5.59-5.59L12 4l-8 8 8 8 1.41-1.41L7.83 13H20v-2z"
                            fill="#000000"
                        />
                    </svg>
                </div>
                <div className="flex flex-col items-start gap-8 w-full mr-2 sm:mr-4 md:mr-8 lg:mr-10">
                    <div className="flex flex-col items-start">
                        <PageTitle value={t('title')} testId={"faq"} />
                        {withHome && (
                            <TertiaryButton testId={'FAQ-Home-Button'} onClick={()=>navigateToDashboardHome(navigate)} title={t('Dashboard:Home.title')}/>
                        )}
                    </div>
                    <div className="w-full">
                        <FAQAccordion />
                    </div>
                </div>
            </div>
        </div>
    );
};
