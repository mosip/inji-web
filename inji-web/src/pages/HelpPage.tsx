import React from "react";
import {HelpAccordion} from "../components/Help/HelpAccordion";
import {NavBar} from "../components/Common/NavBar";
import {useTranslation} from "react-i18next";

export const HelpPage: React.FC = () => {
    const {t} = useTranslation("HelpPage")
    return <React.Fragment>
        <div className={"bg-iw-background"} data-testid="Help-Page-Container">
            <NavBar title={t("title")} search={false} link={"/"}/>
            <div className="container mx-auto mt-8 ">
                <HelpAccordion/>
            </div>
        </div>
    </React.Fragment>
}
