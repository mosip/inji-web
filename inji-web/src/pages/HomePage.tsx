import React, {useEffect} from "react";
import {RequestStatus, useFetch} from "../hooks/useFetch";
import {IntroBox} from "../components/Common/IntroBox";
import {SearchIssuer} from "../components/Issuers/SearchIssuer";
import {IssuersList} from "../components/Issuers/IssuersList";
import {useDispatch} from "react-redux";
import {storeIssuers} from "../redux/reducers/issuersReducer";
import {api, MethodType} from "../utils/api";
import {IssuerObject} from "../types/data";
import {HeaderTile} from "../components/Common/HeaderTile";
import {useTranslation} from "react-i18next";
import {toast} from "react-toastify";

export const HomePage: React.FC = () => {

    const {state, fetchRequest} = useFetch();
    const dispatch = useDispatch();
    const {t} = useTranslation("HomePage");

    useEffect(() => {
        const fetchCall = async () => {
            const response = await fetchRequest(api.fetchIssuers(), MethodType.GET, null);
            const issuers = response?.response?.issuers.filter((issuer: IssuerObject) => issuer.credential_issuer === "Sunbird")
            dispatch(storeIssuers(issuers));
        }
        fetchCall();
    }, [])

    if (state === RequestStatus.ERROR) {
        toast.error(t("errorContent"));
    }

    return <div data-testid="Home-Page-Container">
        <div className="container mx-auto mt-8 px-4 flex-1 flex flex-col">
            <IntroBox/>
            <SearchIssuer state={state} fetchRequest={fetchRequest}/>
        </div>
        <HeaderTile content={t("containerHeading")}/>
        <IssuersList state={state}/>
    </div>

}
