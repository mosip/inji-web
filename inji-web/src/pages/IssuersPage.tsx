import React, {useEffect} from "react";
import {RequestStatus, useFetch} from "../hooks/useFetch";
import {IntroBox} from "../components/Common/IntroBox";
import {SearchIssuer} from "../components/Issuers/SearchIssuer";
import {IssuersList} from "../components/Issuers/IssuersList";
import {useDispatch} from "react-redux";
import {
    storeFilteredIssuers,
    storeIssuers
} from "../redux/reducers/issuersReducer";
import {api} from "../utils/api";
import {ApiRequest, IssuerObject} from "../types/data";
import {useTranslation} from "react-i18next";
import {toast} from "react-toastify";

export const IssuersPage: React.FC<IssuerPageProps> = ({className}) => {
    const {state, fetchRequest} = useFetch();
    const dispatch = useDispatch();
    const {t} = useTranslation("IssuersPage");

    useEffect(() => {
        const fetchCall = async () => {
            const apiRequest: ApiRequest = api.fetchIssuers;
            const response = await fetchRequest(
                apiRequest.url(),
                apiRequest.methodType,
                apiRequest.headers()
            );
            const issuers = response?.response?.issuers.filter(
                (issuer: IssuerObject) => issuer.protocol !== "OTP"
            );
            dispatch(storeFilteredIssuers(issuers));
            dispatch(storeIssuers(issuers));
        };
        fetchCall();
    }, []);

    if (state === RequestStatus.ERROR) {
        toast.error(t("errorContent"));
    }
    return (
        <div data-testid="Home-Page-Container">
            <div className="container mx-auto mt-8 flex flex-col px-4 sm:px-6 md:px-10 lg:px-20">
                <div className={className}>
                    <IntroBox />
                    <SearchIssuer state={state} fetchRequest={fetchRequest} />
                </div>
                <IssuersList state={state} />
            </div>
        </div>
    );
};

type IssuerPageProps = {
    className?: string;
};
