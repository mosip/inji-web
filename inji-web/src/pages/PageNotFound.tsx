import React from "react";
import {useNavigate} from "react-router-dom";
import {useTranslation} from "react-i18next";
import {ROUTES} from "../utils/constants";

export const PageNotFound:React.FC = () => {
    const navigate = useNavigate();
    const location = window.location.pathname
    const {t} = useTranslation("PageNotFound")
    return <div className={"flex justify-center items-center h-full flex-col"}>
        <div>
            <strong>404.</strong>{t("heading")}
        </div>
        <h3 className={"pt-4"}>{t("subHeading", {pathname: window.location.pathname})}</h3>
        <div className={"p-8"}>
          <button
              data-testid="PageNotFound-Home-Button"
              onClick={() => navigate(ROUTES.ROOT)}
              className="text-iw-primary font-bold py-2 px-4 rounded-lg border-2 border-iw-primary">
              {t("navigateButton")}
          </button>
        </div>
    </div>
}
