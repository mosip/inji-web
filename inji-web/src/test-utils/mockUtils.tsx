import React, { ReactElement } from 'react';
import { render, RenderOptions } from '@testing-library/react';
import { Provider } from 'react-redux';
import { BrowserRouter, Route, BrowserRouter as Router, Routes } from 'react-router-dom';
import { reduxStore } from '../redux/reduxStore';

// Mock for storage module
export const mockStorageModule = () => {
    jest.mock('../utils/storage.ts', () => ({
        storage: {
            getItem: jest.fn(),
            setItem: jest.fn(),
            SESSION_INFO: 'SESSION_INFO',
            SELECTED_LANGUAGE: 'selectedLanguage'
        }
    }));
};

// Mock for localStorage
export const mockLocalStorage = () => {
    let store: { [key: string]: string } = {};
  
    const localStorageMock = {
      getItem: jest.fn((key: string) => store[key] || null),
      setItem: jest.fn((key: string, value: string) => {
        store[key] = value;
      })
    };
  
    Object.defineProperty(window, 'localStorage', { 
      value: localStorageMock,
      writable: true
    });
  
    return localStorageMock;
  };

// API Mock
export const mockApi = {
    authorizationRedirectionUrl: 'http://mockurl.com'
};

// Crypto Mock
export const mockCrypto = {
    getRandomValues: (array: Uint32Array) => {
        for (let i = 0; i < array.length; i++) {
            array[i] = 4294967296;
        }
        return array;
    },
    subtle: {} as SubtleCrypto,
    randomUUID: () => '123e4567-e89b-12d3-a456-426614174000'
} as Crypto;

export const mockUseTranslation = () => {
    jest.mock('react-i18next', () => ({
        useTranslation: () => ({
            t: (key: string) => key,
        }),
    }));
};

export const mockUseNavigate = () => {
    const mockNavigate = jest.fn();
    jest.mock('react-router-dom', () => ({
        ...jest.requireActual('react-router-dom'),
        useNavigate: mockNavigate,
    }));
};

export const mockUseSelector = () => {
    jest.mock('react-redux', () => ({
        ...jest.requireActual('react-redux'),
        useSelector: jest.fn(),
    }));
};

export const setMockUseSelectorState = (state: any) => {
    const useSelectorMock = require('react-redux').useSelector;
    useSelectorMock.mockImplementation((selector: any) => selector(state));
};

export const mockUseGetObjectForCurrentLanguage = () => {
    jest.mock('../utils/i18n', () => ({
        getObjectForCurrentLanguage: jest.fn(),
    }));
};

export const wrapUnderRouter = (children: React.ReactNode) => {
    return <Router>{children}</Router>;
};

export const mockUseSearchCredentials = () => {
    jest.mock('../components/Credentials/SearchCredential', () => ({
        SearchCredential: () => <></>,
    }));
 };

export const mockUseDispatch = () => {
    jest.mock('react-redux', () => ({
        ...jest.requireActual('react-redux'),
        useDispatch: jest.fn(),
    }));
};
              
export const mockUseSpinningLoader = () => {
    jest.mock('../components/Common/SpinningLoader', () => ({
        SpinningLoader: () => <></>,
    }));
};

export const mockUseLanguageSelector = () => {
    jest.mock('../components/Common/LanguageSelector', () => ({
        LanguageSelector: () => <></>,
    }));
};

interface RenderWithProviderOptions extends Omit<RenderOptions, 'queries'> {}

export const renderWithProvider = (element: ReactElement, options?: RenderWithProviderOptions) => {
    return render(
        <Provider store={reduxStore}>
            <Router>
                {element}
            </Router>
        </Provider>,
        options
    );
};
    
export const mockUseParams = ()=>{
    jest.mock('react-router-dom', () => ({
        ...jest.requireActual('react-router-dom'),
        useParams: () => ({ issuerId: 'test-issuer-id' }),
      }));   
};

export const mockUseapiObject = () =>{
    jest.mock('../utils/api', () => ({
        api: {
          mimotoHost: 'http://mocked-api-host',
        },
      }));
};

export const mockUseFetch = () =>{
    jest.mock('../hooks/useFetch', () => {
        const RequestStatus = {
          LOADING: 'LOADING',
          DONE: 'DONE',
          ERROR: 'ERROR',
        };
        return {
          useFetch: () => jest.fn(),
          RequestStatus,
        };
      });
};

export const mockUseToast = () =>{
    jest.mock('react-toastify', () => ({
      toast: {
        error: jest.fn(),
      },
    }));
};

export const mockusemisc = ()=>{
    jest.mock('../utils/misc', () => ({
        downloadCredentialPDF: jest.fn(),
        getErrorObject: jest.fn(),
        getTokenRequestBody: jest.fn(),
      }));
};

export const mockusei18n = ()=>{
    jest.mock('react-i18next', () => ({
        useTranslation: () => ({
          t: (key: string) => key,
        }),
        initReactI18next: {
          type: '3rdParty',
          init: jest.fn(),
        },
      }));

};

export const mockWindowLocation = (url: string) => {
    const location = new URL(url) as unknown as Location;
    Object.defineProperty(window, 'location', {
      value: location,
      writable: true,
    });
  };

export const renderWithRouter = (Element: React.ReactElement, { route = '/' } = {}) => {
    window.history.pushState({}, 'Test page', route);
    return render(
      <BrowserRouter>
        <Provider store={reduxStore}>
          <Routes>
            <Route path="*" element={Element} />
          </Routes>
        </Provider>
      </BrowserRouter>
    );
  };

