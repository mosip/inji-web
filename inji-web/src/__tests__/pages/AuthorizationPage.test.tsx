import {AuthorizationPage} from "../../pages/AuthorizationPage";
import {renderWithRouter, mockWindowLocation} from "../../test-utils/mockUtils";

jest.mock("../../utils/api", () => ({
  api: {
    mimotoHost: "http://mocked-api-host"
  }
}));

describe('Testing the Layouts of AuthorizationPage', () => {
  test('Check if it renders properly if there is no error in URL', () => {
    mockWindowLocation('https://api.collab.mosip.net');
    const {asFragment} = renderWithRouter(<AuthorizationPage />);
    expect(asFragment()).toMatchSnapshot();
  });

  test('Check if it shows error messages if error in URL', () => {
    mockWindowLocation('https://api.collab.mosip.net?error=some_error');
    const{asFragment} = renderWithRouter(<AuthorizationPage />);
    expect(asFragment()).toMatchSnapshot();
  });

  afterEach(() => {
    jest.clearAllMocks();
  });
});
