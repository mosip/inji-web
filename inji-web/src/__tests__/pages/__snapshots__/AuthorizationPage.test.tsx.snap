// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing the Layouts of AuthorizationPage Check if it renders properly if there is no error in URL 1`] = `
<DocumentFragment>
  <div>
    <div
      class="flex flex-col justify-center items-center pt-32"
      data-testid="DownloadResult-Outer-Container"
    >
      <div
        class="flex justify-center items-center"
        data-testid="SpinningLoader-Container"
      >
        <div
          aria-busy="true"
          aria-label="oval-loading"
          class="sc-beySPh gZmPOF"
          data-testid="oval-loading"
          role="progressbar"
        >
          <svg
            data-testid="oval-svg"
            height="80"
            stroke="var(--iw-color-spinningLoaderPrimary)"
            viewBox="-20.5 -20.5 43 43"
            width="80"
            xmlns="http://www.w3.org/2000/svg"
          >
            <g
              fill="none"
              fill-rule="evenodd"
            >
              <g
                data-testid="oval-secondary-group"
                stroke-width="3"
                transform="translate(1 1)"
              >
                <circle
                  cx="0"
                  cy="0"
                  r="20"
                  stroke="var(--iw-color-spinningLoaderSecondary)"
                  stroke-opacity=".5"
                  stroke-width="3"
                />
                <path
                  d="M20 0c0-9.94-8.06-20-20-20"
                >
                  <animatetransform
                    attributeName="transform"
                    dur="1s"
                    from="0 0 0"
                    repeatCount="indefinite"
                    to="360 0 0"
                    type="rotate"
                  />
                </path>
              </g>
            </g>
          </svg>
        </div>
      </div>
      <div
        class="my-4"
        data-testid="DownloadResult-Title-Container"
      >
        <p
          class="font-bold"
          data-testid="DownloadResult-Title"
        />
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Testing the Layouts of AuthorizationPage Check if it shows error messages if error in URL 1`] = `
<DocumentFragment>
  <div>
    <div
      class="flex flex-col justify-center items-center pt-32"
      data-testid="DownloadResult-Outer-Container"
    >
      <div
        class="rounded-full p-2 shadow bg-white"
      >
        <div
          class="rounded-full p-8 bg-iw-shieldErrorShadow"
        >
          <svg
            color="var(--iw-color-shieldErrorIcon)"
            data-testid="DownloadResult-Error-ShieldIcon"
            fill="currentColor"
            height="40"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 16 16"
            width="40"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M8 0c-.69 0-1.843.265-2.928.56-1.11.3-2.229.655-2.887.87a1.54 1.54 0 0 0-1.044 1.262c-.596 4.477.787 7.795 2.465 9.99a11.8 11.8 0 0 0 2.517 2.453c.386.273.744.482 1.048.625.28.132.581.24.829.24s.548-.108.829-.24a7 7 0 0 0 1.048-.625 11.8 11.8 0 0 0 2.517-2.453c1.678-2.195 3.061-5.513 2.465-9.99a1.54 1.54 0 0 0-1.044-1.263 63 63 0 0 0-2.887-.87C9.843.266 8.69 0 8 0M6.854 5.146 8 6.293l1.146-1.147a.5.5 0 1 1 .708.708L8.707 7l1.147 1.146a.5.5 0 0 1-.708.708L8 7.707 6.854 8.854a.5.5 0 1 1-.708-.708L7.293 7 6.146 5.854a.5.5 0 1 1 .708-.708"
            />
          </svg>
        </div>
      </div>
      <div
        class="my-4"
        data-testid="DownloadResult-Title-Container"
      >
        <p
          class="font-bold"
          data-testid="DownloadResult-Title"
        >
          error.code.some_error
        </p>
      </div>
      <div
        class="mb-6 px-10 text-center"
        data-testid="DownloadResult-SubTitle-Container"
      >
        <p
          data-testid="DownloadResult-SubTitle"
        >
          error.message.some_error
        </p>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
